lib/go/test/service_test.go

```diff
@@ -401,6 +405,99 @@ func TestContracts(t *testing.T) {
 
 	})
 
+	t.Run("Should check if payer has sufficient balance to execute tx", func(t *testing.T) {
+		// create SmallBalanceContract contract
+		code := []byte(`
+			access(all) contract SmallBalanceContract {
+    			access(all) var value: Int32
+
+    			init() {
+        			self.value = 42
+    			}
+
+				access(all) fun SetValue(new_value: Int32) { 
+					self.value = new_value
+				}
+			}
+    	`)
+		keys := test.AccountKeyGenerator()
+		accKey, accSigner := keys.NewWithSigner()
+		accAddress, err := adapter.CreateAccount(context.Background(), []*flow.AccountKey{accKey}, []sdktemplates.Contract{
+			{
+				Name:   "SmallBalanceContract",
+				Source: string(code),
+			},
+		})
+
+		assert.NoError(t, err)
+		_, err = b.CommitBlock()
+		assert.NoError(t, err)
+
+		// we want to execute some tx so that the payer has less balance than default
+		txCode := []byte(fmt.Sprintf(`
+			import SmallBalanceContract from 0x%s
+			
+			transaction(value: Int32) {
+			  prepare(signer: auth(Storage) &Account) {}
+			
+			  execute {
+			   if value > SmallBalanceContract.value {
+				  SmallBalanceContract.SetValue(new_value: value + 1)
+			   } else {
+				  SmallBalanceContract.SetValue(new_value: value - 1)
+			   }
+			  }
+			}
+		`, accAddress))
+
+		tx := flow.NewTransaction().
+			SetScript(txCode).
+			SetGasLimit(9999).
+			SetProposalKey(b.ServiceKey().Address, b.ServiceKey().Index, b.ServiceKey().SequenceNumber).
+			SetPayer(accAddress).
+			AddAuthorizer(accAddress)
+
+		err = tx.AddArgument(cadence.Int32(15))
+		require.NoError(t, err)
+
+		txRes := signAndSubmit(
+			t, b, tx,
+			[]flow.Address{accAddress},
+			[]crypto.Signer{accSigner},
+			false,
+		)
+		require.True(t, txRes.Succeeded())
```

Comment:
As it is now the transaction is set up with the wrong signers/authorisers. You can try changing this to:

```suggestion
			tx := flow.NewTransaction().
			SetScript(txCode).
			SetComputeLimit(9999).
			SetProposalKey(accAddress, 0, 0).
			SetPayer(accAddress).
			AddAuthorizer(accAddress)

		err = tx.AddArgument(cadence.Int32(15))
		require.NoError(t, err)

		err = tx.SignEnvelope(accAddress, 0, accSigner)
		assert.NoError(t, err)

		// this transaction should fail and be reverted, but the fees will still be paid
		// which will push the balance below the minimum account balance
		// calling VerifyPayerBalanceForTxExecution after this will return false.
		txRes := Submit(t, b, tx, true)
		require.True(t, txRes.Reverted())
```


The key thing to know is that the only way to get the account under the minimum account balance is to send a transaction (while the account is very close to the minimum account balance) that will fail, but some fees will still be deducted, which will result in the account being below minimum account balance.

This will also then mean that `VerifyPayerBalanceForTxExecution` will start returning false for that account.

Reply:
Thank you so much!